presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -1336,6 +1345,77 @@ public Type visitGroupingOperation(GroupingOperation node, StackableAstVisitorCo
             }
         }
 
+        private boolean isDereferenceOrSubscript(Expression node)
+        {
+            return node instanceof DereferenceExpression || node instanceof SubscriptExpression;
+        }
+
+        private boolean isTopMostReference(Expression node, StackableAstVisitorContext<Context> context)
+        {
+            if (!context.getPreviousNode().isPresent()) {
+                return true;
+            }
+            return !isDereferenceOrSubscript((Expression) context.getPreviousNode().get());
+        }
+
+        private void addColumnSubFieldReferences(Expression node, StackableAstVisitorContext<Context> context)
+        {
+            // If expression is nested with multiple dereferences and subscripts, we only look at the topmost one.
+            if (!isTopMostReference(node, context)) {
+                return;
+            }
+            Scope scope = context.getContext().getScope();
+            Expression childNode = node;
+            List<PathElement> columnDereferences = new ArrayList<>();
+            while (true) {
+                if (childNode instanceof SubscriptExpression) {
+                    SubscriptExpression subscriptExpression = (SubscriptExpression) childNode;
+                    childNode = subscriptExpression.getBase();
+                    Type baseType = expressionTypes.get(NodeRef.of(childNode));
+                    if (baseType == null || !(baseType instanceof RowType)) {
+                        continue;
+                    }
```

Comment:
`!(baseType instanceof RowType)` implies `baseType == null`

Reply:
Also remember we have indexed access to fields x.a is same as x[n] if a is the n-th field in the struct.

Reply:
Why? This is for subscript, it could be array/map